app/models/work_package/ancestors.rb

```diff
@@ -0,0 +1,125 @@
+#-- encoding: UTF-8
+#-- copyright
+# OpenProject is a project management system.
+# Copyright (C) 2012-2017 the OpenProject Foundation (OPF)
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2017 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See doc/COPYRIGHT.rdoc for more details.
+#++
+
+module WorkPackage::Ancestors
+  extend ActiveSupport::Concern
+
+  included do
+    attr_accessor :work_package_ancestors
+
+    ##
+    # Retrieve stored eager loaded ancestors
+    # or use awesome_nested_set#ancestors reduced by visibility
+    def visible_ancestors(user)
+      work_package_ancestors.presence || WorkPackage.visible(user).merge(ancestors)
+    end
+  end
+
+  class_methods do
+    def aggregate_ancestors(work_package_ids, user)
+      ::WorkPackage::Ancestors::Aggregator.new(work_package_ids, user).results
+    end
+  end
+
+  ##
+  # Aggregate ancestor data for the given work package IDs.
+  # Ancestors visible to the given user are returned, grouped by each input ID.
+  class Aggregator
+    attr_accessor :user, :ids
+
+    def initialize(work_package_ids, user)
+      @user = user
+      @ids = work_package_ids
+    end
+
+    def results
+      with_work_package_ancestors.group_by(&:leaf_id)
+    end
+
+    private
+
+    def with_work_package_ancestors
+      query = join_ancestors(wp_table)
+
+      WorkPackage.joins(query.join_sources)
+                 .select("#{wp_table.name}.id AS leaf_id")
+                 .select('ancestors.*')
+                 .order('ancestors.lft ASC')
```

Comment:
Out of interest I tried to do the following: 

```
wp1 = WorkPackage.find(x)
wp2 = WorkPackage.find(y)

wp1.ancestors.or(wp2.ancestors)
```

And it works as `ancestors` returns a scope.

So it would also be possible take the work packages and concatenate their respective ancestor-scope via `or`. The resulting SQL however does not look to good as it results in every ancestor SQL being concatenated via `OR`. The DB's query optimizer should make short work of that but I didn't check it and given my experience with MySQL I am not certain it would.

So I am happy to go with your solution which might require more code but results in a more streamlined SQL-Query.